### DISCUSSION TITLE
Your lcov environment

### DISCUSSION THREAD

**Discussion by oberpar** [linux-test-project/lcov#195] [General]
❓ What projects do you use `lcov` for? In what environments do you run it? This information can help us identify what setups should be considered for testing major changes.

- **Project:** <name of the project (if public)>
- **Size:** <size of the project(s) (number of files, functions and lines of code)>
- **Hardware:** <hardware platform (e.g. x86) and typical system characteristics (main memory size, number of CPUs)>
- **OS:** <operating oystem (e.g. Linux)>
- **Language/compiler:** <programming language, compiler type and version (e.g. C/gcc)>

**Comment by oberpar** [comment]
Here's my go:
- **Project:** The Linux kernel 🐧 - specifically the s390-specific part
- **Size:** 5000 files, 120k functions, 1.5M LOC (counting only what `gcc` instrumented for a typical s390 config using kernel v6.2)
- **Hardware:** s390, typically between 4GiB - 1TiB memory, 1-60 CPUs
- **OS:** Linux
- **Language/compiler:** C with current versions of GCC/Clang

**Reply by henry2cox** [reply]
Curious what performance you see from lcov...how long does it take/how much memory to run geninfo and genhtml?
Do you see benefit from parallelizing across cores?  Do you also parallelize across machines (e.g., via LSF)?
Do you see anything we could do, to make it faster?  Would it change or improve your use model, if the tool was X% faster or used Y% less memory?  What is X and Y?

Also curious how the tool/reports are used by the project...as a release criteria?  What criteria do you use?

**Comment by joefradley** [comment]
* **Project:** Android Common Kernel
 * **Size:** ~5000 files, 1.4M LOC (arm64, x86_64 Android configs) 
 * **Hardware:** Various Android devices, real and virtual, arm64 and x86_64
 * **OS:** Linux
 * **Language/compiler:** C with current versions of GCC/Clang

**Reply by henry2cox** [reply]
Same curiousity as with Peter's project :-)

**Reply by joefradley** [reply]
I primarily use lcov to:
 * generate tracefiles from gcov debugfs file dumps
 * merging tracefiles
 * generate html (not as often)

It takes ~1.5 minutes to create a tracefile and I believe about the same to merge two. Because I work with many collections at once this time does become a problem. I ended up running these in parallel using Python scripts.

**Comment by henry2cox** [comment]
Maybe my turn - and sorry to be somewhat vague.

- we use lcov for several different projects, all primarily C/C++ with some Python, Java, and etc. 
- most of the projects are:  C models of various hardware devices, simulators of one form or another, toolchain components (compiler, linker, debugger, etc.), as well as for the product firmware itself.
- almost all use is linux using a mix of gcc (various versions) and llvm (various versions)
- smallest project I'm aware of is about 50K LOC, larger ones are...larger.
- depending on project, we generate both HTML (for developer/manager consumption) and text (for post-processing/tool consumption)
- All the projects I'm aware of use jenkins and some flavor of CI/CD.  Major goal is "sleeping policeman" level automation:  no manual supervision/no intervention unless something bad happens.

I think we may be unrepresentative - even paranoid - about verification because the cost of error in hardware design is so high.

**Comment by jdehaan** [comment]
* **Size:** ~5,000 files, ~300 kLOC
* **OS:** Linux
* **Language/compiler:** C / C++

The project is divided in several suprojects and we use mainly the genhtml capabiltiies to provide users with an easy way to inspect the findings. The generated html is provided via a version control system (a bit awkward but due to the fact it is all decentralized and generated in various places, it was easy to aggregate and provide the information like that)

Beside this at lower level we also use lcov to merge trace files.
